### PR TITLE
Add Create question route and service

### PR DESCRIPTION
**Pull request by XiaoXuxxxx**
### What
Allow users (admin or moderator roles) to create questions and upload the necessary file for questions.

### Why
I don't think letting the admin create questions by editing the database and uploading files directly is a good idea.

### How
First, let me explain the flow for creating question
- insert the data in `question` table in `workspace` database
- insert the data in `question` table in `grading` database
- upload the markdown file to file system
- upload the zipped testcase file to the file system

What have I done in this repo?
- add proto, types for connecting between services 2afa21ef415b78d6927eae30746bd195efa74e08, 6c3d7747c476daaa4d22b9cf0d0e41473cf1d41c
- add route, service, and some repository for creating the question in grading service f851aed48e3b485cb896525d4078a9183fd43249, 30ab63c4cd0bfbbfb3a90a36a7d3a458a25c37ec, c9dd4a8466afe49fe22700fc2672128d41fe2780
- add route, service, and some repository for creating the question in workspace service 84a9769bfaf1512ece11c5d0c781db840ebcf679, dd530fe1c1f53295895f4e5a465efc8f115e096b, 29d9574de625a9a9db13300dc19f7f0af6376737
- add route in gateway service to glue them together 921c1cc7a56f285936e9c39f598456cf8a8b5dcd

### Todo list (if applicable)
- [x] Open my computer
- [x] validate the header from gateway route
- [ ]  upload testcase.zip and markdown file

## Checklist
- [x] Drink juicy beverage when coding ☕️
- [x] Explained the purpose of this PR
- [x] Tested on **Local machine** and verified that there're no visible errors
- [ ] Tested on **Staging server** and verified that there're no visible errors
- [ ] Added or updated test suite for the changes (if applicable)

### CODE REVIEW

**Comment by XiaoXuxxxx** [comment]
เหลืออัพโหลดไฟล์

File: services/workspace/src/services/WorkspaceService.ts [hunk 1]
```diff
@@ -77,4 +83,28 @@ export class WorkspaceService {
...
+      },
+    });
+
+    const updatedPathQuestion = await this.questionRepository.updateQuestionById(
+      createdQuestion.id,
+      { detailPath: `workspaces/${question.workspaceId}/questions/${createdQuestion.id}/question.md` },
+    );
+
+    return updatedPathQuestion;
+  }
```
Comment (riflowth) [inline services/workspace/src/services/WorkspaceService.ts:108]: Why we don't combine a query into a single one?

File: services/gateway/src/controllers/WorkspaceController.ts [hunk 2]
```diff
@@ -112,4 +115,45 @@ export class WorkspaceController {
     return publicQuestions([question], questionSummaries)[0];
   }
 
+  @Post('/:workspaceId/questions')
+  @UseGuards(AuthGuard, WorkspaceOwnerGuard)
```
[the remark this answers is not in the document]
Comment (XiaoXuxxxx) [reply services/gateway/src/controllers/WorkspaceController.ts:119]: lol

File: services/workspace/src/services/WorkspaceService.ts (hunk 1 above)
Comment (XiaoXuxxxx) [reply services/workspace/src/services/WorkspaceService.ts:108]: อยากให้ที่อยู่ของไฟล์มันเกี่ยวโยงกันกับ question id ผมเลยสร้าง question ให้มันพ่น question id ออกมาก่อนอะ

File: services/gateway/src/controllers/WorkspaceController.ts [hunk 3]
```diff
@@ -112,4 +115,45 @@ export class WorkspaceController {
...
+  @UseGuards(FileGuard)
+  public async createQuestion(
+    // TODO: headers validation
```
[the remark this answers is not in the document]
Comment (XiaoXuxxxx) [reply services/gateway/src/controllers/WorkspaceController.ts:122]: thanks kub

File: services/gateway/src/controllers/WorkspaceController.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (XiaoXuxxxx) [reply services/gateway/src/controllers/WorkspaceController.ts:119]: เมา

File: services/gateway/src/controllers/WorkspaceController.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (XiaoXuxxxx) [reply services/gateway/src/controllers/WorkspaceController.ts]: Done kub 4c1f928138b614e2f56ea5906888fe54e0d52262

File: services/gateway/src/controllers/WorkspaceController.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (XiaoXuxxxx) [reply services/gateway/src/controllers/WorkspaceController.ts:120]: Fix in 4c1f928138b614e2f56ea5906888fe54e0d52262 (I will bring FileGuard back when done implementing upload file kub)

File: services/gateway/src/controllers/WorkspaceController.ts [hunk 4]
```diff
@@ -112,4 +115,45 @@ export class WorkspaceController {
...
+    @Param('workspaceId') workspaceId: number,
+    @Headers('question-name') name: string,
+    @Headers('question-description') description: string,
+    @Headers('question-memory-limit') memoryLimit: number,
+    @Headers('question-time-limit') timeLimit: number,
+    @Headers('question-level') level: QuestionLevel,
+    @Headers('question-score') score: number,
+  ): Promise<unknown> {
```
[the remark this answers is not in the document]
Comment (XiaoXuxxxx) [reply services/gateway/src/controllers/WorkspaceController.ts]: done kub 35b6874d7b194d4a632d39b76bbbe715b56e1a0e

**Comment by riflowth** [comment]
@XiaoXuxxxx Thanks for hard-working! Let continue the implementation in the main repository

**Pull request merged**